src/libraries/System.Runtime/tests/System.Reflection.Tests/TypeTests.cs

```diff
@@ -0,0 +1,258 @@
+        [InlineData(typeof(string*))]
+        [InlineData(typeof(void*))]
+        [InlineData(typeof(char**))]
+        [InlineData(typeof(TypeTests*))]
+        public static void Namespace_Pointers_ReturnsElementTypeNamespace(Type type)
+        {
+            Type elementType = type.GetElementType();
+            Assert.Equal(elementType.Namespace, type.Namespace);
+        }
+
+        [Fact]
+        public static void Namespace_FunctionPointers_ReturnsNull()
+        {
+            Assert.Null(typeof(delegate*<void>).Namespace);
+            Assert.Null(typeof(delegate*<void>*).Namespace);
+            Assert.Null(typeof(delegate*<void>**).Namespace);
+        }
+
+        [Fact]
+        public static void Namespace_ByRefTypes_ReturnsElementTypeNamespace()
+        {
+            Type byRefInt = typeof(int).MakeByRefType();
+            Type byRefString = typeof(string).MakeByRefType();
+            Type byRefCustom = typeof(TypeTests).MakeByRefType();
+
+            Assert.Equal("System", byRefInt.Namespace);
+            Assert.Equal("System", byRefString.Namespace);
+            Assert.Equal("System.Reflection.Tests", byRefCustom.Namespace);
+        }
+
+        [Theory]
+        [InlineData(typeof(Collections.Generic.Dictionary<int, string>), "System.Collections.Generic")]
+        [InlineData(typeof(Collections.Generic.List<int>), "System.Collections.Generic")]
+        [InlineData(typeof(Collections.Generic.KeyValuePair<string, int>), "System.Collections.Generic")]
+        [InlineData(typeof(Nullable<int>), "System")]
+        [InlineData(typeof(Func<int, string>), "System")]
+        [InlineData(typeof(Action<int>), "System")]
+        public static void Namespace_ConstructedGenericTypes_ReturnsGenericTypeDefinitionNamespace(Type type, string expectedNamespace)
+        {
+            Assert.Equal(expectedNamespace, type.Namespace);
+        }
+
+        [Fact]
+        public static void Namespace_NestedTypes_ReturnsDeclaringTypeNamespace()
+        {
+            // Use existing nested types from the test assembly
+            Assert.Equal("System.Reflection.Tests", typeof(TI_BaseClass.PublicNestedClass1).Namespace);
+            Assert.Equal("System.Reflection.Tests", typeof(TI_BaseClass.PublicNestedClass2).Namespace);
+            Assert.Equal("System.Reflection.Tests", typeof(TI_SubClass.PublicNestedClass1).Namespace);
+            Assert.Equal("System.Reflection.Tests", typeof(MultipleNestedClass.Nest1).Namespace);
+            Assert.Equal("System.Reflection.Tests", typeof(MultipleNestedClass.Nest1.Nest2).Namespace);
+            Assert.Equal("System.Reflection.Tests", typeof(MultipleNestedClass.Nest1.Nest2.Nest3).Namespace);
+        }
+
+        [Fact]
+        public static void Namespace_TopLevelTypeWithNoNamespace_ReturnsNull()
+        {
+            // Create a simple dynamic assembly with a type that has no namespace
+            var assemblyBuilder = System.Reflection.Emit.AssemblyBuilder.DefineDynamicAssembly(
```

Comment:
Reflection.Emit needs ConditionalFact with IsReflectionEmitSupported.

This should not need Reflection.Emit in the first place. Top level types in C# have empty namespace, so it should be possible to write the test type in C#.